pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/MultipleOrEqualitiesToInClauseFilterQueryTreeOptimizer.java

```diff
@@ -18,163 +18,221 @@
- * 3</code> gets turned to <code>a IN (1, 2, 3)</code>.
+ * The optimizer looks at the FilterQueryTree to convert
+ *
+ * (1) multiple OR equality predicates into a single IN predicte
+ * e.g COL = 1 OR COl = 2 OR COL = 3 is rewritten as COL IN (1, 2, 3)
+ *
+ * (2) single value IN predicate into equality predicate
+ * e.g COL IN (1) is rewritten as COL = 1
+ *
+ * (3) Deduplicates the predicate values for IN.
+ *
+ * These rewrites are also done for child subtrees (handling nested nature)
  */
 public class MultipleOrEqualitiesToInClauseFilterQueryTreeOptimizer extends FilterQueryTreeOptimizer {
   @Override
   public FilterQueryTree optimize(FilterQueryOptimizerRequest request) {
-    return optimize(request.getFilterQueryTree(), null);
+    return optimize(request.getFilterQueryTree());
   }
 
-  private FilterQueryTree optimize(FilterQueryTree filterQueryTree, FilterQueryTree parent) {
+  private FilterQueryTree optimize(FilterQueryTree filterQueryTree) {
     if (filterQueryTree.getOperator() == FilterOperator.OR) {
-      Map<String, Set<String>> columnToValues = new HashMap<>();
-      List<FilterQueryTree> nonEqualityOperators = new ArrayList<>();
+      // Root operator is OR
+      return optimizeTreeRootedAtOR(filterQueryTree);
+    } else if (filterQueryTree.getOperator() == FilterOperator.AND) {
+      // Root operator is AND
+      return optimizeTreeRootedAtAND(filterQueryTree);
+    } else {
+      // if we are here, it means the query has a single predicate
+      // (just one column with one or more predicate values)
+      // a = 1
+      // a IN (1)
+      // a > 7
+      // a IN (1, 2, 3, 4)
+      // .... anything single predicate
+      return singlePredicateOptimization(filterQueryTree);
+    }
+  }
 
-      // Collect all equality/in values and non-equality operators
-      boolean containsDuplicates = collectChildOperators(filterQueryTree, columnToValues, nonEqualityOperators);
+  private FilterQueryTree singlePredicateOptimization(FilterQueryTree filterQuery) {
+    // rewrite single predicate IN as EQUALITY, for remaining cases this optimizer is a NO-OP
+    if (filterQuery.getOperator() == FilterOperator.IN && filterQuery.getValue().size() == 1) {
+      filterQuery.setOperator(FilterOperator.EQUALITY);
+    }
+    return filterQuery;
+  }
 
-      // If we have at least one column to return
-      if (!columnToValues.isEmpty()) {
-        // We can eliminate the OR node if there is only one column with multiple values
-        if (columnToValues.size() == 1 && nonEqualityOperators.isEmpty()) {
-          Map.Entry<String, Set<String>> columnAndValues = columnToValues.entrySet().iterator().next();
+  private static class PredicateInfo {
+    Set<String> _predicateValues;
+    boolean _multipleOccurences;
+    boolean _inPredicate;
 
-          return buildFilterQueryTreeForColumnAndValues(columnAndValues);
-        }
+    PredicateInfo(Set<String> predicateValues, boolean multipleOccurences, boolean inPredicate) {
```

Comment:
Why do we need these extra info besides predicateValues?

Reply:
In addition to predicateValues, we need a boolean rewritePredicate to indicate if that predicate needs to be rewritten

Reply:
All the information is now self-contained in this data structure